File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/allowances/LinearAllowance.java
package fr.sncf.osrd.envelope_sim.allowances;

import static fr.sncf.osrd.envelope_sim.TrainPhysicsIntegrator.POSITION_EPSILON;

import fr.sncf.osrd.envelope.*;
import fr.sncf.osrd.envelope.part.EnvelopePart;
import fr.sncf.osrd.envelope_sim.EnvelopeProfile;
import fr.sncf.osrd.envelope_sim.EnvelopeSimContext;
import fr.sncf.osrd.envelope_sim.PhysicsRollingStock;
import fr.sncf.osrd.envelope_sim.allowances.utils.AllowanceRange;
import java.util.*;

public class LinearAllowance extends AbstractAllowanceWithRanges {

    /** Constructor */
    public LinearAllowance(
            double beginPos,
            double endPos,
            double capacitySpeedLimit,
            List<AllowanceRange> ranges
    ) {
        super(beginPos, endPos, capacitySpeedLimit, ranges);
    }

    /** Compute the initial low bound for the binary search */
    @Override
    protected double computeInitialLowBound(Envelope envelopeSection) {
        return capacitySpeedLimit;
    }

    /** Compute the initial high bound for the binary search */
    @Override
    protected double computeInitialHighBound(Envelope envelopeSection, PhysicsRollingStock rollingStock) {
        return envelopeSection.getMaxSpeed();
    }

    /** Compute the core of linear allowance algorithm.
     *  This algorithm consists of a ratio that scales speeds */
    @Override
    protected Envelope computeCore(Envelope coreBase, EnvelopeSimContext context, double maxSpeed) {
        var ratio = maxSpeed / coreBase.getMaxSpeed();
        return scaleEnvelope(coreBase, ratio);
    }

    /** Scale an envelope, new speed = old speed * ratio */
    public static Envelope scaleEnvelope(Envelope envelope, double ratio) {
        var builder = new EnvelopeBuilder();
        for (var part : envelope)
            builder.addPart(scalePart(part, ratio));
        return builder.build();
    }

    /** Scale a single part, new speed = old speed * ratio */
    private static EnvelopePart scalePart(EnvelopePart part, double ratio) {
        var positions = new ArrayList<Double>();
        var speeds = new ArrayList<Double>();
        for (int i = 0; i < part.pointCount(); i++) {
            if (i > 0 && Math.abs(positions.get(i - 1) - part.getPointPos(i)) < POSITION_EPSILON) {
                // Remove points that are too close to one another, could cause problems with scaling down
                // The first one is removed to keep the same values at the end of the envelope
                positions.remove(positions.size() - 1);
                speeds.remove(speeds.size() - 1);
            }
            positions.add(part.getPointPos(i));
            speeds.add(part.getPointSpeed(i) * ratio);
        }
        var attr = part.getAttr(EnvelopeProfile.class);
        var attrs = List.<EnvelopeAttr>of();
        if (attr != null)
            attrs = List.of(attr);
        return EnvelopePart.generateTimes(
                attrs,
                positions.stream().mapToDouble(x -> x).toArray(),
                speeds.stream().mapToDouble(x -> x).toArray()
        );
    }
}

File: core/envelope-sim/src/main/kotlin/fr/sncf/osrd/envelope_sim/allowances/LinearAllowance.kt
package fr.sncf.osrd.envelope_sim.allowances

import fr.sncf.osrd.envelope.Envelope
import fr.sncf.osrd.envelope.EnvelopeBuilder
import fr.sncf.osrd.envelope.part.EnvelopePart
import fr.sncf.osrd.envelope.part.EnvelopePart.Companion.generateTimes
import fr.sncf.osrd.envelope_sim.EnvelopeProfile
import fr.sncf.osrd.envelope_sim.EnvelopeSimContext
import fr.sncf.osrd.envelope_sim.PhysicsRollingStock
import fr.sncf.osrd.utils.SelfTypeHolder

class LinearAllowance(
    beginPos: Double,
    endPos: Double,
    capacitySpeedLimit: Double,
    ranges: List<AllowanceRange>,
) : AbstractAllowanceWithRanges(beginPos, endPos, capacitySpeedLimit, ranges.toMutableList()) {
    /** Compute the initial low bound for the binary search */
    override fun computeInitialLowBound(envelopeSection: Envelope): Double {
        return capacitySpeedLimit
    }

    /** Compute the initial high bound for the binary search */
    override fun computeInitialHighBound(
        envelopeSection: Envelope,
        rollingStock: PhysicsRollingStock,
    ): Double {
        return envelopeSection.maxSpeed
    }

    /**
     * Compute the core of linear allowance algorithm. This algorithm consists of a ratio that
     * scales speeds
     */
    override fun computeCore(
        base: Envelope,
        context: EnvelopeSimContext,
        speedCap: Double, // Max speed
    ): Envelope {
        val ratio = speedCap / base.maxSpeed
        return scaleEnvelope(base, ratio)
    }

    companion object {
        /** Scale an envelope, new speed = old speed * ratio */
        fun scaleEnvelope(envelope: Envelope, ratio: Double): Envelope {
            val builder = EnvelopeBuilder()
            for (part in envelope) builder.addPart(scalePart(part, ratio))
            return builder.build()
        }

        /** Scale a single part, new speed = old speed * ratio */
        private fun scalePart(part: EnvelopePart, ratio: Double): EnvelopePart {
            val positions = part.getPositions().map { obj: Double -> obj }.toDoubleArray()
            val speeds = part.getSpeeds()
            val scaledSpeeds = speeds.map { it * ratio }.toDoubleArray()
            val attr = part.getAttr(EnvelopeProfile::class.java)
            var attrs = listOf<SelfTypeHolder>()
            if (attr != null) attrs = listOf(attr)
            return generateTimes(attrs, positions, scaledSpeeds)
        }
    }
}
